src/Orleans.DurableJobs/ShardExecutor.cs

```diff
@@ -107,33 +107,61 @@ private async Task RunJobAsync(
     {
         await Task.CompletedTask.ConfigureAwait(ConfigureAwaitOptions.ContinueOnCapturedContext | ConfigureAwaitOptions.ForceYielding);
 
+        Exception? failureException = null;
+
         try
         {
             LogExecutingJob(_logger, jobContext.Job.Id, jobContext.Job.Name, jobContext.Job.TargetGrainId, jobContext.Job.DueTime);
 
             var target = _grainFactory.GetGrain<IDurableJobReceiverExtension>(jobContext.Job.TargetGrainId);
 
-            await target.DeliverDurableJobAsync(jobContext, cancellationToken);
-            await shard.RemoveJobAsync(jobContext.Job.Id, cancellationToken);
+            var result = await target.DeliverDurableJobAsync(jobContext, cancellationToken);
 
-            LogJobExecutedSuccessfully(_logger, jobContext.Job.Id, jobContext.Job.Name);
-        }
-        catch (Exception ex) when (ex is not TaskCanceledException)
-        {
-            LogErrorExecutingJob(_logger, ex, jobContext.Job.Id);
-            var retryTime = _options.ShouldRetry(jobContext, ex);
-            if (retryTime is not null)
+            // Handle the result based on status
+            while (result.Status == DurableJobRunStatus.PollAfter)
             {
-                LogRetryingJob(_logger, jobContext.Job.Id, jobContext.Job.Name, retryTime.Value, jobContext.DequeueCount);
-                await shard.RetryJobLaterAsync(jobContext, retryTime.Value, cancellationToken);
+                // Enter polling loop
+                LogPollingJob(_logger, jobContext.Job.Id, jobContext.Job.Name, result.PollAfterDelay!.Value);
```

Comment:
If you want to get fancy, you can have an `IsPending` property with `[MemberNotNullWhen(true, nameof(PollAfterDelay)]` and then you can remove the ! here. Similar for the other properties.